### PR TITLE
Add element locator on displaced mesh and other related changes

### LINKED ISSUE #6962: Make ElementLocator/Info work with displaced mesh
### Description of the enhancement or error report
Recent we added pressure bc on XFEM cut surfaces but we only considered the undeformed mesh case. For deformed mesh, we need to modify some base class of ElementLocator/Info such that it can work for large deformation problem.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
should have no impacts on existing codes.

### PR DESCRIPTION
**Pull request by jiangwen84**
This PR does not include test but test will be added in PR #6862

Closes #6962

### CODE REVIEW

**Comment by permcody** [comment]
hmmm..... Should I merge more code from @jiangwen84 or wait until the current xfem code is clean? 😉

**Comment by jiangwen84** [comment]
I will work on the memory issue right after this. :-)

**Comment by permcody** [comment]
Hey @jiangwen84, how's the memory leak hunting going? Anything we can do to help? I've updated the suppressions file a few times today to remove spurious errors. I think only the XFEM code still has errors.

**Comment by jiangwen84** [comment]
I think Valgrind detects multiple memory leaking places. Some of the information is not easy to understand. For example, from the valgrind info 

```
==4263== 448,885 (128 direct, 448,757 indirect) bytes in 1 blocks are definitely lost in loss record 1,889 of 1,889
==4263==    at 0x4C2B1C7: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4263==    by 0x53430AA: XFEMElementPairLocator::reinit() (new_allocator.h:104)
==4263==    by 0x810758A: GeometricSearchData::reinit() (GeometricSearchData.C:138)
==4263==    by 0x8353C8A: FEProblem::meshChanged() (FEProblem.C:3758)
==4263==    by 0x832E86E: FEProblem::updateMeshXFEM() (FEProblem.C:3726)
==4263==    by 0x834EE9C: FEProblem::initialSetup() (FEProblem.C:493)
==4263==    by 0x8960276: Transient::init() (Transient.C:191)
==4263==    by 0x8522C30: MooseApp::executeExecutioner() (MooseApp.C:496)
==4263==    by 0x852768C: MooseApp::run() (MooseApp.C:653)
==4263==    by 0x40BEDE: main (main.C:37)
```

I was looking into ```XFEMElementPairLocator::reinit() ``` function, but in which I think we did not use ```new``` to dynamically allocate memory. Could you let me know what could be wrong with this code. I believe we have other similar codes like this that might cause the memory leak.

**Comment by friedmud** [comment]
I took a look at that code and I can't see it either.  The only thing I can
think of is that the whole XFEMElementPairLocator is being lost / not
destroyed somehow.

Derek

**Comment by jiangwen84** [comment]
Thanks Derek. Actually we use ```MooseSharedPointer``` for ```XFEMElementPairLocator```. In ```XFEMAction.C```, ```MooseSharedPointer<XFEMElementPairLocator> new_xfem_epl (new XFEMElementPairLocator(xfem, 0))```;  :-)

**Comment by friedmud** [comment]
I saw that... so I am definitely perplexed. That should be handling the
cleanup.

Maybe someone else can spot something I'm missing..,

**Comment by permcody** [comment]
There are other dynamic allocations happening throughout that code. Those
allocated objects are being stored in maps but I also think they can move
around a bit making tracing the origins very difficult.

Fixing these leaks should be top priority. I'd like a clean code base
before we move to C++11. Thanks

**Comment by jiangwen84** [comment]
@permcody I found that in ```XFEM``` tests, the destructor of ```FEProblem``` and ```XFEM``` is never called. I think it might relate to the fact that ```FEProblem``` contains ```XFEM smart pointer``` and ```XFEM``` also contains ```FEProblem smart pointer``` and this causes that they can not be freed automatically. Do you think if it makes any sense to you? Thanks.

**Comment by permcody** [comment]
Yes that makes perfect sense.

C++11 handles that with "weak pointers". You should be able to release one of the pointers in the ring manually for now. Then the other will get cleaned up once the ring is broken.

**Comment by jiangwen84** [comment]
@permcody Could you change the label from Ready to Merge to Under Development?  

I think Ben's PR#7002 should be merged first and I will rework on this PR based on that.

**Comment by jiangwen84** [comment]
@bwspenc could you review this code? :-)

File: modules/xfem/src/geomsearch/XFEMElementPairLocator.C [hunk 1]
```diff
@@ -41,5 +46,41 @@ XFEMElementPairLocator::reinit()
...
+    // elem2
+    _xfem->getXFEMIntersectionInfo(elem2, plane_id, normal, intersectionPoints, _use_displaced_mesh);
+
+    if (intersectionPoints.size() == 2)
+      _xfem->getXFEMqRuleOnLine(intersectionPoints, q_points, weights);
+    else if (intersectionPoints.size() > 2)
+      _xfem->getXFEMqRuleOnSurface(intersectionPoints, q_points, weights);
+
+    ElementPairInfo new_elem_info2(elem2, q_points, weights, normal);
+    _element_pair_info.insert(std::pair<const Elem*, ElementPairInfo>(elem2, new_elem_info2));
```
Comment (bwspenc) [inline modules/xfem/src/geomsearch/XFEMElementPairLocator.C]: My only major comment on this is that it seems a little odd that you're creating a second `ElementPairInfo` object for the second element. That's not really a problem, but I guess based on the naming of that object (it's for an element pair, after all), I'd expect to see a single instance of that object for each element/element constraint, and would expect that object to contain the information for both elements involved in the constraint. Following a pattern like that might make it easier to extend this to recursively split elements down the road as well.

File: framework/src/geomsearch/GeometricSearchData.C [hunk 2]
```diff
@@ -60,6 +60,15 @@ GeometricSearchData::update(GeometricSearchType type)
           it != _slave_to_qslave.end();
           ++it)
         generateQuadratureNodes(it->first, it->second);
+
+      // reinit on displaced mesh before update
+      for (std::map<unsigned int, MooseSharedPointer<ElementPairLocator> >::iterator epl_it = _element_pair_locators.begin();
+           epl_it != _element_pair_locators.end(); ++epl_it)
+      {
+        ElementPairLocator & epl = *epl_it->second;
+
```
Comment (bwspenc) [inline framework/src/geomsearch/GeometricSearchData.C]: I'd get rid of this empty line.

File: framework/src/geomsearch/GeometricSearchData.C [hunk 3]
```diff
@@ -98,6 +107,14 @@ GeometricSearchData::update(GeometricSearchType type)
       pl->detectPenetration();
     }
   }
+
+  for (std::map<unsigned int, MooseSharedPointer<ElementPairLocator> >::iterator epl_it = _element_pair_locators.begin();
+       epl_it != _element_pair_locators.end(); ++epl_it)
+  {
+    ElementPairLocator & epl = *epl_it->second;
+
```
Comment (bwspenc) [inline framework/src/geomsearch/GeometricSearchData.C]: This empty line too.

File: framework/src/geomsearch/GeometricSearchData.C [hunk 4]
```diff
@@ -135,6 +152,7 @@ GeometricSearchData::reinit()
        epl_it != _element_pair_locators.end(); ++epl_it)
   {
     ElementPairLocator & epl = *epl_it->second;
+
```
Comment (bwspenc) [inline framework/src/geomsearch/GeometricSearchData.C]: And this one.

File: modules/xfem/src/geomsearch/XFEMElementPairLocator.C (hunk 1 above)
Comment (jiangwen84) [inline modules/xfem/src/geomsearch/XFEMElementPairLocator.C]: I see you point. Do you think if I should add a class/struct inside ```ElementPairInfo``` to hold the quadrature info for each element? Or if you have any better ideas?

Comment (bwspenc) [inline modules/xfem/src/geomsearch/XFEMElementPairLocator.C]: I think I'd just add a second set of data for the second element in there. Maybe something like:  `_neighbor_elem`, `_neighbor_constraint_q_point`, `_neighbor_constraint_JxW`, `_neighbor_normal`.

**Comment by jiangwen84** [comment]
@bwspenc Could you take a look at this again?

File: framework/include/geomsearch/ElementPairInfo.h [hunk 5]
```diff
@@ -39,16 +39,31 @@ class ElementPairInfo
 public:
 
   ElementPairInfo(const Elem * elem,
+                  const Elem * pair_elem,
```
Comment (bwspenc) [inline framework/include/geomsearch/ElementPairInfo.h]: @jiangwen84 I think this all looks good, except that I think the naming here is a little confusing. How about `elem1`, `elem2`, `elem1_constraint_q_point`, `elem2_constraint_q_point`, etc?

Comment (jiangwen84) [inline framework/include/geomsearch/ElementPairInfo.h]: No problem. That will be an easy fix. :)

**Comment by bwspenc** [comment]
This looks good to me now.

**Comment by permcody** [comment]
@jiangwen84 - It looks like we are going to have to start running Valgrind for all these XFEM updates. I know this is complex code, but we need to keep the code base clean. This is causing new Valgrind errors:
https://www.moosebuild.org/job/24154/

**Comment by jiangwen84** [comment]
@permcody  I am sorry about that. I will fix it and make sure it can pass Valgrind test.

**Pull request merged**